samcli/lib/intrinsic_resolver/intrinsics_symbol_table.py

```diff
@@ -0,0 +1,424 @@
+            This is a clean way of specifying common attributes across all types.
+            The value can either be a function of the form string or (logical_id) => string
+            {
+                "Ref": lambda p,r: "",
+                "Arn:": arn_resolver
+            }
+        """
+        self.logical_id_translator = logical_id_translator or {}
+
+        self.parameters = parameters or {}
+        self.resources = resources or {}
+
+        self.default_type_resolver = default_type_resolver or {
+            "AWS::ApiGateway::RestApi": {
+                "RootResourceId": "/"  # It usually used as a reference to the parent id of the RestApi,
+            }
+        }
+
+        self.common_attribute_resolver = common_attribute_resolver or {
+            "Ref": lambda logical_id: logical_id,
+            "Arn": self.arn_resolver
+        }
+
+        self.default_pseudo_resolver = {
+            IntrinsicsSymbolTable.AWS_ACCOUNT_ID: self.handle_pseudo_account_id,
+            IntrinsicsSymbolTable.AWS_NOTIFICATION_ARN: self.handle_pseudo_notification_arn,
+            IntrinsicsSymbolTable.AWS_PARTITION: self.handle_pseudo_partition,
+            IntrinsicsSymbolTable.AWS_REGION: self.handle_pseudo_region,
+            IntrinsicsSymbolTable.AWS_STACK_ID: self.handle_pseudo_stack_id,
+            IntrinsicsSymbolTable.AWS_STACK_NAME: self.handle_pseudo_stack_name,
+            IntrinsicsSymbolTable.AWS_NOVALUE: self.handle_pseudo_no_value,
+            IntrinsicsSymbolTable.AWS_URL_PREFIX: self.handle_pseudo_url_prefix,
+        }
+
+    def resolve_symbols(self, logical_id, resource_attribute, ignore_errors=False):
+        """
+        This function resolves all the symbols given a logical id and a resource_attribute for Fn::GetAtt and Ref.
+        This boils Ref into a type of Fn:GetAtt to simplify the implementation.
+        For example:
+            {"Ref": "AWS::REGION"} => resolve_symbols("AWS::REGION", "REF")
+            {"Fn::GetAtt": ["logical_id", "attribute_type"] => resolve_symbols(logical_id, attribute_type)
+
+
+        First pseudo types are checked. If item is present in the logical_id_translator it is returned.
+        Otherwise, it falls back to the default_pseudo_resolver
+
+        Then the default_type_resolver is checked, which has common attributes and functions for each types.
+        Then the common_attribute_resolver is run, which has functions that are common for each attribute.
+        Parameters
+        -----------
+        logical_id: str
+            The logical id of the resource in question or a pseudo type.
+        resource_attribute: str
+            The resource attribute of the resource in question or Ref for psuedo types.
+        ignore_errors: bool
+            An optional flags to not return errors. This used in sub
+
+        Return
+        -------
+        This resolves the attribute
+        """
+        # pylint: disable-msg=too-many-return-statements
```

Comment:
This is an indication that we should split this function out or find a better way to manage the control flows to reduce return statements (in some sane way).